django/db/models/constraints.py

```diff
@@ -642,30 +642,34 @@ def validate(self, model, instance, exclude=None, using=DEFAULT_DB_ALIAS):
                     meta=model._meta, exclude=exclude
                 ).items()
             }
-            expressions = []
+            conditions = []
             for expr in self.expressions:
                 if hasattr(expr, "get_expression_for_validation"):
                     expr = expr.get_expression_for_validation()
-                expressions.append(Exact(expr, expr.replace_expressions(replacements)))
-            queryset = queryset.filter(*expressions)
+                lhs, rhs = expr, expr.replace_expressions(replacements)
+                condition = Exact(lhs, rhs)
+                if self.nulls_distinct is False:
+                    condition = Q(condition) | Q(IsNull(lhs, True), IsNull(rhs, True))
```

Comment:
I guess we could eventually have a `NotDistinctFrom(Lookup)` lookup that would use `{lhs} IS NOT DISTINCT FROM {rhs}` on backend that supports it and emulate it otherwise (e.g. `<=>` [on MySQL](https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_equal-to), fallback to `OR ...`). This should be good for now though since this is our first requirement for an internal usage.

Reply:
Yes that sounds good. 
Just to expand on what I was trying to communicate [here](https://github.com/django/django/pull/18356#issuecomment-2217977675), I was expecting to find some code that would have looked like
```python
...
if self.nulls_distinct is not False:
    condition = Q(condition) & Q(IsNull(lhs, False))
```
and was surprised that it isn't necessary during the unique check. But I see everything works correctly, this is just for my understanding.

Reply:
After re-reading your comment on the other ticket I understand this a little better now, thanks.